### PR TITLE
SDT-64 Create Consolidated jar file during Gradle build

### PR DESCRIPTION
**Pull request by markdathornehmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/SDT-64

### Change description ###

During Gradle build process, make sure it is creating consolidated jar file with all the dependencies and all sub modules included in the generated jar.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -187,15 +187,34 @@ allprojects {
...
   dependencies {
+    implementation files(
+      'consumers/build/libs/civil-sdt-consumers-0.0.1-SNAPSHOT.jar',
+      'dao/build/libs/civil-sdt-dao-0.0.1-SNAPSHOT.jar',
+      'dao-api/build/libs/civil-sdt-dao-api-0.0.1-SNAPSHOT.jar',
+      'domain/build/libs/civil-sdt-domain-0.0.1-SNAPSHOT.jar',
+      'handlers/build/libs/civil-sdt-handlers-0.0.1-SNAPSHOT.jar',
+      'handlers-api/build/libs/civil-sdt-handlers-api-0.0.1-SNAPSHOT.jar',
+      'interceptors/build/libs/civil-sdt-interceptors-0.0.1-SNAPSHOT.jar',
+      'producers/build/libs/civil-sdt-producers-0.0.1-SNAPSHOT.jar',
```
Comment (paulridout) [inline build.gradle:206]: The original SDT build produces two release files: producers.war (SDT) and producers-comx.war (SDT commissioning).  The producers.war file is built from the producers project and references the other projects JAR files (consumers.jar, dao.jar etc.).  This is different in that it builds one archive that references all the project JARs including producers.  Will this work?

Comment (markdathornehmcts) [reply build.gradle:206]: Good query. Not sure but I don't see why not - all the classes are available while still within their own module jar. Unfortunately we don't have anything that can run. just yet. Producers commissioning hasn't been looked at (yet).

Comment (paulridout) [reply build.gradle:206]: OK.  We'll go with what we've got for the moment and revisit it later should the need arise.

File: build.gradle [hunk 2]
```diff
@@ -187,15 +187,34 @@ allprojects {
...
+      'producers-api/build/libs/civil-sdt-producers-api-0.0.1-SNAPSHOT.jar',
+      'services/build/libs/civil-sdt-services-0.0.1-SNAPSHOT.jar',
+      'services-api/build/libs/civil-sdt-services-api-0.0.1-SNAPSHOT.jar',
+      'transformers/build/libs/civil-sdt-transformers-0.0.1-SNAPSHOT.jar',
+      'utils/build/libs/civil-sdt-utils-0.0.1-SNAPSHOT.jar',
+      'validators/build/libs/civil-sdt-validators-0.0.1-SNAPSHOT.jar')
+
```
[the remark this answers is not in the document]
Comment (paulridout) [reply build.gradle:213]: To be addressed later as per other review comment.

**Review by paulridout — APPROVED** [review]
Changes look OK.

File: build.gradle [hunk 3]
```diff
@@ -187,15 +187,34 @@ allprojects {
...
+      'consumers/build/libs/civil-sdt-consumers-0.0.1-SNAPSHOT.jar',
+      'dao/build/libs/civil-sdt-dao-0.0.1-SNAPSHOT.jar',
+      'dao-api/build/libs/civil-sdt-dao-api-0.0.1-SNAPSHOT.jar',
+      'domain/build/libs/civil-sdt-domain-0.0.1-SNAPSHOT.jar',
+      'handlers/build/libs/civil-sdt-handlers-0.0.1-SNAPSHOT.jar',
+      'handlers-api/build/libs/civil-sdt-handlers-api-0.0.1-SNAPSHOT.jar',
+      'interceptors/build/libs/civil-sdt-interceptors-0.0.1-SNAPSHOT.jar',
+      'producers/build/libs/civil-sdt-producers-0.0.1-SNAPSHOT.jar',
+      'producers-api/build/libs/civil-sdt-producers-api-0.0.1-SNAPSHOT.jar',
+      'services/build/libs/civil-sdt-services-0.0.1-SNAPSHOT.jar',
```
Comment (kiran-yenigala-hmcts) [inline build.gradle:208]: bootJar {
  manifest {
    attributes 'Implementation-Title': project.name,
            'Implementation-Version': project.version
  }
  duplicatesStrategy = DuplicatesStrategy.EXCLUDE
  archiveName = 'civil-sdt.jar'
  from {
    subprojects.collect { it.sourceSets.main.output }
  }
}

File: build.gradle [hunk 4]
```diff
@@ -480,6 +499,12 @@ subprojects { subproject ->
     failFast = true
   }
 
+  tasks.jar {
+    enabled = true
+    // Remove `plain` postfix from jar file name
+    archiveClassifier.set("")
```
Comment (kiran-yenigala-hmcts) [inline build.gradle:505]: if we are happy with above solution, then we can remove this

File: validators/build.gradle [hunk 5]
```diff
@@ -21,7 +21,6 @@ dependencies {
 }
 
 jar {
-    enabled = true
```
Comment (kiran-yenigala-hmcts) [inline validators/build.gradle:24]: we can revert all these changes back to normal if above solution works, or remove this completely

**Review by kiran-yenigala-hmcts — CHANGES_REQUESTED** [review]
Please see the comments about bootJar suggession

File: build.gradle (hunk 3 above)
Comment (paulridout) [reply build.gradle:208]: The producers-commissioning and sdt-consumer-sample projects will need to be excluded from this as they aren't part of SDT itself.

File: build.gradle [hunk 6]
```diff
@@ -567,15 +567,19 @@ idea {
...
 // this allows ./gradlew run
 bootJar {
-
   manifest {
     attributes 'Implementation-Title': project.name,
             'Implementation-Version': project.version
   }
+  duplicatesStrategy = DuplicatesStrategy.EXCLUDE
   archiveName = 'civil-sdt.jar'
+  from {
```
Comment (kiran-yenigala-hmcts) [inline build.gradle:576]: if we have this one inplace, I don't think below is needed

  "consumers/build/libs/civil-sdt-consumers-${version}.jar",
      "dao/build/libs/civil-sdt-dao-${version}.jar",
      "dao-api/build/libs/civil-sdt-dao-api-${version}.jar",
      "domain/build/libs/civil-sdt-domain-${version}.jar",

Comment (kiran-yenigala-hmcts) [reply build.gradle:576]: can you please check this one

**Review by kiran-yenigala-hmcts — APPROVED** [review]

**Pull request merged**